Library/Homebrew/extend/ENV/std.rb

```diff
@@ -57,15 +57,20 @@ def setup_build_environment(formula = nil)
         # See https://github.com/Linuxbrew/linuxbrew/issues/841
         prepend_path "LD_LIBRARY_PATH", formula.opt_lib
         prepend_create_path "LD_LIBRARY_PATH", formula.prefix
-        prepend "LD_LIBRARY_PATH", formula.lib, File::PATH_SEPARATOR
+        prepend_path "LD_LIBRARY_PATH", formula.lib
       end
 
       # Set the search path for header files.
       prepend_path "CPATH", HOMEBREW_PREFIX/"include"
       # Set the dynamic linker and library search path.
-      append "LDFLAGS", "-Wl,--dynamic-linker=#{HOMEBREW_PREFIX}/lib/ld.so -Wl,-rpath,#{HOMEBREW_PREFIX}/lib"
       prepend_path "LIBRARY_PATH", HOMEBREW_PREFIX/"lib"
+      prepend "LDFLAGS", "-Wl,-rpath=#{HOMEBREW_PREFIX}/lib"
       prepend_path "LD_RUN_PATH", HOMEBREW_PREFIX/"lib"
+      unless formula.nil?
+        prepend "LDFLAGS", "-Wl,-rpath=#{formula.lib}"
```

Comment:
To be clear, the issue affects only keg-only formulae (because their libraries aren't symlinked into `$HOMEBREW_PREFIX/lib`), but we're adding the lib directory for all formulae, right?

I can think of one possible consequence of that, which is that if you have two conflicting formulae that both install a shared library and you force one of them with `brew link --overwrite`, the library from this formula will still be preferred even if it's overwritten. I'm not sure if that's a good thing or a bad thing; I'm leaning towards *good thing*, but some users might find it surprising.